src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.BoxedTypes.cs

```diff
@@ -89,6 +89,19 @@ public MethodDesc GetSpecialUnboxingThunk(MethodDesc targetMethod, ModuleDesc ow
             return thunk;
         }
 
+        public MethodDesc GetUnspecialUnboxingThunk(MethodDesc targetMethod, ModuleDesc ownerModuleOfThunk)
+        {
+            TypeDesc owningType = targetMethod.OwningType;
+            Debug.Assert(owningType.IsValueType);
+
+            var owningTypeDefinition = (MetadataType)owningType.GetTypeDefinition();
```

Comment:
This is correctly going to the type definition, but we also need to do the inverse to specialize back before returning the result.

Consider we need an unboxing thunk for method `Foo<SomeType>.Bar` - this should end up creating an unboxing thunk for `Boxed_Foo<T>.Bar` that we then specialize to `Boxed_Foo<SomeType>.Bar`.

See how `GetSpecialUnboxingThunk` achieves that with `GetTypicalMethodDefinition()` and `GetMethodForInstantiatedType`. The special unboxing thunk only deals with generics, so it just asserts `owningType != owningTypeDefinition`, but in your case you'll have an `if` to check if we need to do that.